src/commandExecutor.ts

```diff
@@ -177,16 +179,32 @@ const CommandExecutor = async () => {
       era,
     }
     validateWitnessing(signingParameters)
-    const txWitnesses = await cryptoProvider.witnessTx(signingParameters, args.changeOutputKeyFileData)
-    if (txWitnesses.length < args.outFiles.length) {
-      // eslint-disable-next-line no-console,max-len
-      console.log(`Warning! ${args.outFiles.length} output file(s) specified, but only ${txWitnesses.length} witness(es) generated. The remaining output file(s) were not written to.`)
+    const {
+      byronWitnesses, shelleyWitnesses,
+    } = await cryptoProvider.witnessTx(signingParameters, args.changeOutputKeyFileData)
+    const txWitnesses = [...byronWitnesses, ...shelleyWitnesses]
+
+    const txWitnessOutputs: WitnessOutput[] = []
+    for (let i = 0; i < args.hwSigningFileData.length; i += 1) {
+      const signingFilePath = args.hwSigningFileData[i].path
+      const witness = txWitnesses.find((w) => pathEquals(w.path, signingFilePath))
+      if (witness !== undefined) {
+        txWitnessOutputs.push(constructTxWitnessOutput(era, witness))
+      } else {
+        // eslint-disable-next-line no-console,max-len
+        console.log(`Warning! A superfluous HW signing file specified (${i + 1} of ${args.hwSigningFileData.length}), the witness was not created.`)
+      }
     }
-    if (txWitnesses.length > args.outFiles.length) {
+    if (txWitnessOutputs.length > args.outFiles.length) {
       throw Error(Errors.NotEnoughOutFilesError)
     }
-    for (let i = 0; i < txWitnesses.length; i += 1) {
-      write(args.outFiles[i], constructTxWitnessOutput(era, txWitnesses[i]))
+    for (let i = 0; i < args.outFiles.length; i += 1) {
+      if (i < txWitnessOutputs.length) {
+        write(args.outFiles[i], txWitnessOutputs[i])
```

Comment:
I'd rename the `write` function to something more specific, e.g. writeOutputData (similarly to writeCbor).

Reply:
Renamed in https://github.com/vacuumlabs/cardano-hw-cli/pull/122/commits/d39f9bb088252f4daa75320e17d77cc26f3c6fd8.